### PR TITLE
fix: Relationship.relationshipType is a MetadataIdentifier

### PR DESCRIPTION
**Pull request by teleivo**
## This PR

* migrate Relationship.relationshipType to MetadataIdentifier
* map RelationshipType in DuplicateRelationshipSupplier so we do not get LazyInitExceptions
* map key/invertedKey in RelationshipMapper
* Add idScheme related fields to Relationship/TypeMappers

## Background on changes in [DHIS2-12563](https://jira.dhis2.org/browse/DHIS2-12563)

* metadata identifier fields in org.hisp.dhis.tracker.domain will gradually be made idScheme aware.
* tracker view (JSON import/export) and domain models are split since users only specify idSchemes in the query parameters (not in the body) and only on import. The metadata identifier fields in the user-facing import/export JSON are strings containing only the identifier without idScheme.

all metadata identifiers in the tracker domain
* Attribute.attribute
* DataValue.dataElement
* Event.programStage :heavy_check_mark:
* Event.attributeOptionCombo :heavy_check_mark:
* Event.attributeCategoryOptions :heavy_check_mark:
* TrackedEntity.trackedEntityType :heavy_check_mark:
* Relationship.relationshipType :heavy_check_mark:
* {Enrollment :heavy_check_mark:,Event :heavy_check_mark:}.program
* {Enrollment :heavy_check_mark:,Event :heavy_check_mark:,TrackedEntity :heavy_check_mark:}.orgUnit

will become idScheme aware.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/RelationshipPreheatKeySupport.java [hunk 1]
```diff
@@ -55,8 +55,9 @@ public static boolean hasRelationshipKey( Relationship relationship )
 
     public static RelationshipKey getRelationshipKey( Relationship relationship )
     {
+        // TODO() is the type always stored using UIDs?
         return getRelationshipKey(
-            relationship.getRelationshipType(),
+            relationship.getRelationshipType().getIdentifier(),
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/RelationshipPreheatKeySupport.java:60]: We are collecting the keys from the payload and check in the DB if we found any relationship with the same key.
So we need the key to be a MetadataIdentifier

**Review by lucaCambi77 — APPROVED** [review]

**Review by enricocolasante — COMMENTED** [review]
We probably need to discuss about the RelationshipKey. I am not sure as it is now it will work. Definitely it was not working before.
We have a in the DB a String representing the key where we have the relationshipType but we do not actually know the idScheme there in the String, if we search for it using a different idScheme we will not find anything even though we were supposed to.

We are always using the UID to save the key in the DB, we should somehow use the UID to create the key in the preheat.
And I think we have the same problem with the identifiers of the relationshipItems from and to.

We really need to understand if it makes sense to have a string representing the key in the DB now that we are dealing with idSchemes

**Review by enricocolasante — CHANGES_REQUESTED** [review]
Put as a change request so it will not auto merge

**Comment by teleivo** [comment]
> And I think we have the same problem with the identifiers of the relationshipItems from and to.

interesting and 😢 since I am very close to finishing and tired of idSchemes 😂 I was wondering if we have good tests for this area. The whole String key being a concatenation of identifiers/parsing it again when retrieving is a bit 😨 

relationshipItems from and to are TEI, enrollment, events which are data so only support UIDs. That should be fine. Or am I missing something?

**Comment by enricocolasante** [comment]
> > And I think we have the same problem with the identifiers of the relationshipItems from and to.
> 
> interesting and 😢 since I am very close to finishing and tired of idSchemes 😂 I was wondering if we have good tests for this area. The whole String key being a concatenation of identifiers/parsing it again when retrieving is a bit 😨
> 
> relationshipItems from and to are TEI, enrollment, events which are data so only support UIDs. That should be fine. Or am I missing something?

You are right, relationshipItems are not a problem, only relationshipType.
I am not sure how much we have covered with the tests.

**Comment by stian-sandvold** [comment]
> > And I think we have the same problem with the identifiers of the relationshipItems from and to.
> 
> interesting and 😢 since I am very close to finishing and tired of idSchemes 😂 I was wondering if we have good tests for this area. The whole String key being a concatenation of identifiers/parsing it again when retrieving is a bit 😨
> 
> relationshipItems from and to are TEI, enrollment, events which are data so only support UIDs. That should be fine. Or am I missing something?

As Enrico mentions, the key, in the database, will not work as it is now, since the type identifier might be something other than uid. So we would need to either get the UID before we preheat relationships, or rewrite how we fetch them.

I agree the key thing in the database is 😨, but the problem is that when the tables are normalized, the joining and lookup is actually super slow. This helped a bit, but it's still struggelig as reported by Gintare. If you have a good alternative to the lookup that is performing similar or better, that also allows us to normalize the tables, I'm all for it. But until we have that in place, I'm not sure we can merge this PR, as it would break the relationship preheat, which means relationships would not be validated by uniqueness anymore.

With the change @enricocolasante is making to relationships (UID being a primary way of identifying them), this logic might be ripe for an update regardless, since we can exclude any lookup using the key for relationships with a UID that exists; Ideally, if we preheat the entities in the relationships (Get their primary ids) that might speed up the lookup now that I think about it. In any case, we'll have to figure this out before we proceed with the PR as it is

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**